server/routes/api/protein.py

```diff
@@ -13,22 +13,411 @@
+            cross_targets.append(target_id)
+        else:
+            expansion_targets.append(target_id)
+    return {
+        'expansion_targets': expansion_targets,
+        'cross_targets': cross_targets
+    }
+
+
+@bp.route('/protein-protein-interaction/', methods=['POST'])
+def protein_protein_interaction():
+    '''
+    Retrieves graph data for protein-protein interaction graph via degree-bounded BFS
+
+    Request params:
+    - proteinDcid: the DCID of the page protein (e.g. 'bio/P53_HUMAN')
+    - scoreThreshold: interaction score threshold above which to display an edge between two interacting proteins
+    - numInteractors: we take {numInteractors} top-scoring expansion links of each node during BFS
+    - depth: the maximum distance of any returned node from the center protein
+
+    Returns: A dict of form
+    {
+        'nodeDataNested': [[...depth 0 node dicts], [...depth 1 node dicts], ...]
+        'linkDataNested': [[...depth 0 link dicts], [...depth 1 link dicts], ...]
+    }
+    The only depth 0 node is the center protein and since we do not support self-interactions at the moment,
+    there are no depth-0 links, so for depth m, the above can be simplified to
+    {
+        'nodeDataNested': [[<center protein>], [...depth 1 nodes], ..., [...depth m nodes]]
+        'linkDataNested': [[], [...depth 1 links], ..., [...depth m links]]
+    }.
+
+    Each node dict has form
+    {
+        'id': id of form '{name}_{species}', e.g. 'P53_HUMAN',
+        'name': name, e.g. 'P53',
+        'species': species, e.g. 'HUMAN',
+        'depth': currently the zero-indexed iteration of breadth-bounded BFS protein was added
+        TODO: update depth so that it equals the distance of protein to center
+        (Note the two are the same when numInteractors = infinity)
+    }
+
+    Each link dict has form
+    {
+        'source': id of source protein,
+        'target': id of target protein,
+        'score': IntactMi score of interaction
+    }
+
+    TODO: example
+    '''
+
+    try:
+        center_protein_dcid = request.json['proteinDcid']
+        score_threshold = request.json['scoreThreshold']
+        max_interactors = request.json['maxInteractors']
+        max_depth = request.json['maxDepth']
+
+        if type(score_threshold) not in [int, float]:
+            raise TypeError('scoreThreshold must be a number')
+        if type(max_interactors) != int:
+            raise TypeError("maxInteractors must be an integer")
+        if type(max_depth) != int:
+            raise TypeError("maxDepth must be an integer")
+        if type(center_protein_dcid) != str:
+            raise TypeError("proteinDcid must be a string")
+
+    except KeyError as key_error:
+        return f'Missing request parameter {key_error}', BAD_REQUEST_CODE
+
+    except TypeError as type_error:
+        return f'Incorrect request parameter type: {type_error}', BAD_REQUEST_CODE
+
+    for param, limits in LIMITS.items():
```

Comment:
From the conversation yesterday, doesn't seem like we need these limits anymore?

Reply:
These may have to be reintroduced in the future but I think currently the graph has 26k proteins with interaction data and 77,000 interactions so it would not be terrible for the user to load everything.